### DISCUSSION TITLE
In game plugin installation and updating

### DISCUSSION THREAD

**Discussion by RisingLeaf** [endless-sky/endless-sky#9058] [Ideas]
## Introduction
I recently started working on a system to make installing and updating plugins for ES much easier. I want both things to happen inside of ES and not in an external launcher.
What I got until now is a very basic implementation of a system that reads a file in the format:
"plugin name" "github url"
and then displays these plugins as a list, and the player can select plugins he wants to install.
This comes with the downside of having to restart the game for actually using the plugins, but I am sure that would not be too hard to fix. The UI for the whole thing is also a bit messy (because ES UI coding is a pain).
https://github.com/RisingLeaf/endless-sky/tree/plugin-install

But that's not what I want to talk about. I thought before I invest more time into coding this I would just ask if they way I am approaching this is actually the way we would like to go for this.

## Explanation of the implementation
Right now I am using a neat library called libgit2, which allows you to do basically any git commands as c++ code. It is full of useful tools and definitely the right tool for what I want to do. It also works on linux, mac and windows.
What I am thinking is this:
Most of our plugin authors have their plugins stored on GitHub, so why not use that fact to our advantage? Instead of having to create another hosting system for plugins we can just use GitHub (or any other git server) to host the plugins and all we have to provide is a list of trusted plugins, Hecters plugin archive can work as a base for that.

## The Question
Do you people think this way is the right one to go?
(also is ideas the right category?)

**Comment by mOctave** [comment]
Ideas is the right category probably, but I think the officially managed plugin directory should be the primary source of plugins in-game. That's not to say that letting people download any plugin in Hecter's archive is a bad idea, but I know for certain that I have plugins on GitHub (and in that archive) that are probably not ones that people would (or should) want to download in their current state. Even having dropdowns or something, one for relatively polished plugins and one—or more, divided by category—for everything in the archive would be nice.

**Reply by lumbar527** [reply]
All plugins should probably move to the official plugins directory. Zitchas says that [here](https://github.com/endless-sky/endless-sky/discussions/7928#discussioncomment-5276230), and I agree with it and 1010todd's reply about tagging

**Reply by Hecter94** [reply]
Any official plugin list should be opt-in, whereas the archive is opt-out.

**Reply by MasterOfGrey** [reply]
> Any official plugin list should be opt-in, whereas the archive is opt-out.

Agreed.

However, if there was some sort of tag we could use with the archive that would opt those specific plugins in to be either made accessible, or duplicated to somewhere else, or what have you, that would be ideal to my mind.

Opt-in tagging would be separate to any tagging as to the development status of the plugin.

**Reply by RisingLeaf** [reply]
I do agree that it should be opt-in
but the official plugin registry seems a bit outdated, so if we could have some tag in the plugin archive that lets us filter those plugins and use a script to compose a list of them, that would indeed be great.

**Reply by MCOfficer** [reply]
Why does it seem outdated? It only has those plug-ins authors choose to add, but most of them are kept uptodate semi-automatically. Updates are usually merged within an hour if i'm awake, or once I wake up.

**Reply by RisingLeaf** [reply]
oh? seems I had the wrong impression on that point

**Reply by RisingLeaf** [reply]
In that case it would probably be best to just read the plugins.json file?

**Reply by MCOfficer** [reply]
sure, or the index can generate any other machine-readable format if that's easier.

**Reply by RisingLeaf** [reply]
I dont know if we want to add a json loader to ES, so maybe a simple txt file would be better?
Using the standard node style:
```txt
plugin "name"
    url "url"
    description "description"
    author "author"
```
I think that would be good, because the loader for that style is already in the code

**Reply by MCOfficer** [reply]
Doable, it would require implementing it ourselves, which brings in a bunch of baggage like quoting and such. There are header-only json parsers like https://github.com/sheredom/json.h, if that's acceptable on the ES side i'd much prefer using a well-defined and tested format like json.

**Reply by RisingLeaf** [reply]
if that works for people, json would probably be easier to handle yes

**Reply by RisingLeaf** [reply]
although, i think using a full fledged library might be better

**Reply by mOctave** [reply]
JSON is great for machines, good for programmers, and (in my opinion) absolutely horrible for non-coders compared with something like YAML or ES' node format. As long as no one has any reason to edit the JSON file who doesn't have a thorough understanding of JSON, I'm fine with it. Otherwise, though, I'd personally much prefer ES' node format since it is more consistent with everything else and much easier to understand.

**Reply by Zireael07** [reply]
I agree that JSON is awful for non-coders who would likely prefer YAML. The question is, whether any non-coders would need to touch the json?

**Reply by MCOfficer** [reply]
Not really. The [index files](https://github.com/endless-sky/endless-sky-plugins/tree/master/manifests) are maintained (by humans & automatically) as YAML. It periodically converts them to [machine-readable formats](https://github.com/endless-sky/endless-sky-plugins/tree/master/generated), i.e. JSON, which applications can read. If ES uses that JSON, no human needs to touch it.

**Reply by RisingLeaf** [reply]
One thing that I would like to know then would be how we provide that json, do we include it in every release or do we update it every time the game opens?

**Reply by MCOfficer** [reply]
given how often some *cough midnight cough* plugins update, i'd much prefer downloading it. We need to download things anyways to install the plugins.

**Reply by RisingLeaf** [reply]
actually that would not matter, updating a plugin can be done with libgit2 alone, there just would be no way to add plugins

**Reply by RisingLeaf** [reply]
[plugin-install](https://github.com/RisingLeaf/endless-sky/tree/plugin-install)
I implemented loading the [plugins.json](https://github.com/endless-sky/endless-sky-plugins/blob/master/generated/plugins.json) into my branch and I think it is a really simple approach, which is a good thing.

**Comment by RisingLeaf** [comment]
https://github.com/endless-sky/endless-sky/pull/9068 linking this here

**Comment by RisingLeaf** [comment]
Discussion should be on the PR now
